### PR TITLE
Added ability to pass vector into parameterised query

### LINKED ISSUE #303: Add support for using arrays in parameterized queries
The bigrquery package is a great interface; thanks for all who have contributed to its development! It would be very helpful if the package could be extended to handle arrays in parameterized queries, [a la their API documentation](https://cloud.google.com/bigquery/docs/parameterized-queries#bigquery_query_params_arrays-python). It seems that currently passing in array parameters is not supported in bigrquery.

I often find myself wanting to subset data from my BigQuery tables based on a set of IDs generated from another dataset that lives elsewhere. Have tried a couple ways of passing an array into `bq_dataset_query()` but gotten the error message: "Second argument of IN UNNEST must be an array but was STRING at..." 

Here's an example of what I've been trying. The commented out code show variations I've tried adding.

```R
library(dplyr)
library(bigrquery)
library(jsonlite)

project <- "bigquery-public-data"
dataset <- "un_sdg"

q <- "
  SELECT geoareaname, goal, count(goal) as n
  FROM `un_sdg.indicators`
  WHERE geoareaname IN UNNEST(@countries)
  GROUP BY geoareaname, goal
"

params <- c("Belgium", "Brazil", "Bulgaria") # %>% array() %>% toJSON()

d <- bq_dataset_query(
  x = paste0(project, ".", dataset),
  query = q,
  parameters = list(countries = params)
) %>%
  bq_table_download()

glimpse(d)
```

### PR DESCRIPTION
**Pull request by byapparov**
closes #303 

test demonstrates this feature:

```R
  ds <- as_bq_dataset("bigquery-public-data.moon_phases")
  job <- bq_perform_query(
    "SELECT COUNT(*) cnt FROM moon_phases WHERE phase IN UNNEST(@phases)",
    parameters = list(phases = c("First Quarter", "Last Quarter")),
    billing = bq_test_project(),
    default_dataset = ds
  )
  job <- bq_job_wait(job)
  job_tb <- bq_job_table(job)

  df <- bq_table_download(job_tb)

  # -- query to get expected result: 9894
  # SELECT COUNT(*) FROM `bigquery-public-data.moon_phases.moon_phases`
  # WHERE phase IN UNNEST(ARRAY<STRING>["First Quarter", "Last Quarter"])
  expect_equal(df$cnt, 9894)

```
this is the json intercepted from execution of the code above:

```json
{
  "configuration": {
    "query": {
      "query": "SELECT COUNT(*) cnt FROM moon_phases WHERE phase IN UNNEST(@phases)",
      "useLegacySql": false,
      "priority": "INTERACTIVE",
      "queryParameters": [
        {
          "name": "phases",
          "parameterType": {
            "type": "ARRAY",
            "arrayType": {
              "type": "STRING"
            }
          },
          "parameterValue": {
            "arrayValues": [
              {
                "value": "First Quarter"
              },
              {
                "value": "Last Quarter"
              }
            ]
          }
        }
      ],
      "defaultDataset": {
        "projectId": "bigquery-public-data",
        "datasetId": "moon_phases"
      }
    }
  }
} 
```

### CODE REVIEW

File: tests/testthat/test-bq-perform.R [hunk 1]
```diff
@@ -74,8 +74,8 @@ test_that("bq_perform_query creates job that succeeds", {
 test_that("can supply parameters", {
   ds <- as_bq_dataset("bigquery-public-data.moon_phases")
   job <- bq_perform_query(
-    "SELECT * FROM moon_phases WHERE peak_datetime = @date",
-    parameters = list(date = bq_param("1889-07-28 00:00:00", "DATETIME")),
+    "SELECT * FROM moon_phases WHERE peak_datetime = CAST(@date AS DATETIME)",
```
Comment (byapparov) [inline tests/testthat/test-bq-perform.R]: This was done to make test independent from internals of the package. e.g. I could not run this test in IDE.

**Comment by byapparov** [comment]
Hi @hadley / @jennybc, do you mind integrating this PR?

**Comment by jennybc** [comment]
I am also experimenting with a change to the codecov configuration in this PR. Many tests won't run for a PR, because a PR can't use our encrypted tokens. This renders the coverage report for PRs almost useless, hence my proposal. The `threshold` is the maximum drop in coverage that is acceptable, so that's why it's 100% and not 0%.

BTW this PR passes tests when I run them locally, with access to a token.

**Comment by byapparov** [comment]
It would make sense to have some branching strategy where contributors could create new branch inside r-dbi/bigrquery so the tests are executed or have instructions on how to make tests run in forks?

It would be nice to see all tests passing inside the PR

**Comment by hadley** [comment]
I don't think we can merge this right away; it at least has styling problems, and I would need to get my head into bigrquery space to properly analyse it. I don't have the time to do that right now.

**Comment by hadley** [comment]
There is no way to set up a strategy whereby you can run your code with our credentials, because then you could modify the code to expose the credentials (which matters here because there's a credit card associated with the billing account).

**Comment by byapparov** [comment]
@hadley styling I am happy to fix with more details from you.

do you lint the code?

**Comment by jennybc** [comment]
We use the styler package.

**Comment by byapparov** [comment]
@jennybc project does not pass `styler::style_pkg()` call, while my changes do.

Do I need to provide some specific style specification into the call?

**Comment by byapparov** [comment]
@jennybc is there any chance to get feedback on this PR while you are working on this package. I am not sure how priorities work on your side. It's important for applications we do, e.g. any shinyapp accessing bq would benefit from this.

It also important as a security feature. Not sure if DML injection is possible on bigquery, so I would block any code with UI and paste0 over SQL template from going to production.

Would really appreciate any feedback on PR so we could integrate it.

**Comment by hadley** [comment]
As I have already said, I don't have time to properly analyse it. I am sceptical that it's correct, because in general switch on the length of something is an anti-pattern. Your tests also don't look correct because they should be testing `as_json()`.

**Comment by byapparov** [comment]
Hi @hadley,

 I have done what I could to improve it:

* Testing. My test was absolutely fine as the acceptance test. There were no unit tests for the `as_json` to build on. I have added unit test now, which can serve as documentation on how params work. 

* Switching on length - it possibly could be better to isolate that logic somewhere, e.g. Class Factory, which will require additional separation of `bq_param` class into `bq_param_single`, `bq_param_array`, `bq_param_struct` for example. On top of this you could ask users of the package to provide parameter type inside the  call.  In any case check for the value length will be still required. 

I appreciate you are busy, hope this will help you to review it without spending too much time on API internals. 

Thanks
Bulat

File: R/bq-param.R [hunk 2]
```diff
@@ -28,15 +30,35 @@ as_bq_params <- function(x) {
...
-  out <- vector("list", length(x))
-  for (i in seq_along(x)) {
-    out[[i]] <- list(
-      name = names(x)[[i]],
-      parameterType = list(type = unbox(x[[i]]$type)),
-      parameterValue = list(value = x[[i]]$value)
+#' Parameters of STRUCT type are not supported
+#'
+#' @noRd
+as_json_bq_param <- function(x, name) {
```
Comment (byapparov) [inline R/bq-param.R]: @hadley here I could not do `as_json.bq_param` because `as_json` only takes one argument. Alternative to this would be to set name of parameter inside of its structure. 

e.g. 
```r
bq_param(value, type, name)
```

**Comment by byapparov** [comment]
this will not work in cases where vector is of length one. 

I will see if I can find a solution for it. It seems that user of the package will have to specify type of parameter in the call as there is no way to check if parameter is used inside the `UNNEST()` and it does not work with scalars.

**Comment by byapparov** [comment]
@hadley I have updated the code per my comment on 30 Jun. 

Key changes are:

- new classes introduced `bq_param_scalar` and `bq_param_array`
- users should be encouraged to be explicit on which one to use.

I did not make any breaking changes so as before parameter name is defined via list key value. 
This is not the cleanest solution from the OOP point of view, but I prefer that from the user point of view.

**Old inexplicit**
```r
job <- bq_perform_query(
    query_template,
    parameters = list(
      year = 2000L,
      phases =c("First Quarter", "Last Quarter")
    ),
    billing = bq_test_project(),
    default_dataset = ds
  )
```

**New explicit**

```r
job <- bq_perform_query(
    query_template,
    parameters = list(
      year = 2000L,
      phases = bq_param_array(c("First Quarter", "Last Quarter"))
    ),
    billing = bq_test_project(),
    default_dataset = ds
  )
```

In both cases name is injected afterwards into the parameter structure. 

The following will also work, but is not forced on the user.

```r
job <- bq_perform_query(
    query_template,
    parameters = list(
     bq_param_scalar(2000L, name = "year"),
     bq_param_array(
       c("First Quarter", "Last Quarter"), 
       name = "phases"
      )
    ),
    billing = bq_test_project(),
    default_dataset = ds
  )
```

**Comment by byapparov** [comment]
Hi @hadley, is there any chance of this moving forward?

**Comment by hadley** [comment]
It will move forward when I am next working on bigrquery. Reviewing it when I'm not working on bigrquery is likely to end up with a suboptimal outcome because I don't remember how all the pieces fit together.

**Comment by byapparov** [comment]
Hi @hadley,

I would like to offer my help as an autonomous contributor to `bigrquery`.

I can do it because:
- I am happy to dedicate several hours a week
- I use BigQuery and R every day
- I know BigQuery API and organisation of `bigrquery` package
- I write necessary and sufficient tests first
- I respect conventions set in projects that I contribute to
- I have experience in helping other developers, you can see my stackoverflow profile [here](https://stackoverflow.com/users/599911/bulat-reinstate-monica)

Few reasons why `bigrquery` needs more contributors:
- It is an extremely valuable and popular package for the R community
- Package is getting stale, it is considerably behind Py libraries
- Bugs and security issues take too long to resolve
- Forking is a painful alternative

Here are few R packages where I have contributed in the past:
- https://github.com/dirkschumacher/rcbc
- https://github.com/nealrichardson/httptest
- https://github.com/byapparov/googleAnalyticsR

Let me know what you think.

Best Regards
Bulat

File: NEWS.md [hunk 3]
```diff
@@ -36,6 +39,8 @@ gargle and rlang are newly Imported.
 # bigrquery 1.1.1
 
 * Fix test failure with dbplyr 1.4.0.
+* `bq_perform_query()` can now execute parameterised query with parameters
```
Comment (hadley) [inline NEWS.md:42]: Can you please take a pass through this PR and clean up changes that don't belong? i.e. this news bullet appears twice

File: R/bq-param.R [hunk 4]
```diff
@@ -23,20 +63,39 @@ bq_params <- function(x) {
...
 
-  out <- vector("list", length(x))
-  for (i in seq_along(x)) {
-    out[[i]] <- list(
-      name = names(x)[[i]],
-      parameterType = list(type = unbox(x[[i]]$type)),
-      parameterValue = list(value = x[[i]]$value)
-    )
-  }
+#' Parameters of STRUCT type are not supported
```
Comment (hadley) [inline R/bq-param.R:78]: This doesn't seem like it should be a roxygen2 comment

File: R/bq-request.R [hunk 5]
```diff
@@ -118,7 +118,7 @@ bq_delete <- function(url, ..., query = NULL, token = bq_token()) {
 
 #' @importFrom httr POST add_headers config
 bq_post <- function(url, body, ..., query = NULL, token = bq_token()) {
-  json <- jsonlite::toJSON(body, pretty = TRUE, auto_unbox = TRUE)
+  json <- jsonlite::toJSON(body, pretty = TRUE, auto_unbox = TRUE, na = "null")
```
Comment (hadley) [inline R/bq-request.R:121]: Why does this need to change?

File: R/bq-stream.R [hunk 6]
```diff
@@ -0,0 +1,89 @@
+#' Streams data from data.from into BigQUery
```
Comment (hadley) [inline R/bq-stream.R:1]: This doesn't seem related to parameterised queries?

File: codecov.yml [hunk 7]
```diff
@@ -5,8 +5,8 @@ coverage:
     project:
       default:
         target: auto
-        threshold: 1%
+        threshold: 100%
```
Comment (hadley) [inline codecov.yml:8]: Please revert changes to this file

File: tests/testthat/test-bq-perform.R (hunk 1 above)
Comment (hadley) [reply tests/testthat/test-bq-perform.R:77]: You should be exploring interactively after using `load_all()` so you can access non-exported functions.

File: tests/testthat/test-bq-perform.R [hunk 8]
```diff
@@ -85,4 +85,116 @@ test_that("can supply parameters", {
   df <- bq_table_download(job_tb)
   expect_equal(nrow(df), 1)
   expect_equal(df$phase, "New Moon")
+
+})
+
+test_that("can supply parameters as array for IN statement", {
+  ds <- as_bq_dataset("bigquery-public-data.moon_phases")
+  job <- bq_perform_query(
```
Comment (hadley) [inline tests/testthat/test-bq-perform.R:93]: Can you simplify these test cases? I think you should be able to test array parameters without querying a dataset (i.e. like the examples in https://cloud.google.com/bigquery/docs/reference/standard-sql/arrays)

File: R/bq-request.R (hunk 5 above)
Comment (byapparov) [reply R/bq-request.R:121]: Without this JSON would be wrong and (I think they were wrongly defaulted to `string` here and would fail the tests)

File: R/bq-stream.R (hunk 6 above)
Comment (byapparov) [reply R/bq-stream.R:1]: Yes, this was due to merges on my fork. I can try to clean it up if your prefer.

File: codecov.yml (hunk 7 above)
Comment (byapparov) [reply codecov.yml:8]: this was changed by Jennifer (@jennybc) for experiments I suppose, happy to change back if not needed here

File: tests/testthat/test-bq-perform.R (hunk 8 above)
Comment (byapparov) [reply tests/testthat/test-bq-perform.R]: I have updated tests that were added by me. Do you want all tests to be updated in the same way?

File: R/bq-param.R (hunk 4 above)
Comment (byapparov) [reply R/bq-param.R:80]: I have added details, I am not sure where is the best place to put it. may be it should be moved downstream to `bq_perform_query`

File: R/bq-request.R (hunk 5 above)
Comment (byapparov) [reply R/bq-request.R:121]: Technically `na = "null"` is default option, but for some reason it did not work at some point. 

I think it makes sense to have it set explicitly here in any case.

File: tests/testthat/test-bq-perform.R (hunk 1 above)
Comment (byapparov) [reply tests/testthat/test-bq-perform.R:78]: I think in this case based on the fact that test is for `bq_perform_query`, it should not access anything that user (developer) cannot access.

File: R/bq-stream.R (hunk 6 above)
Comment (hadley) [reply R/bq-stream.R]: Yes please; the cleaner your PR is, the easiest it is for me to review.

**Comment by byapparov** [comment]
@hadley I have responded to all your comments now.

File: NEWS.md [hunk 9]
```diff
@@ -1,10 +1,10 @@
 # bigrquery (development version)
 
+* `bq_perform_query()` can now execute parameterised query with parameters 
+  of ARRAY type (@byapparov, #303).
+
 * Error messages inlude hints for common problems (@deflaux, #353).
 
-* `bq_perform_upload()` works once again (#361). It seems like the generated
```
Comment (hadley) [inline NEWS.md:5]: I think you deleted this accidentally?

File: tests/testthat/test-bq-perform.R (hunk 1 above)
Comment (hadley) [reply tests/testthat/test-bq-perform.R:78]: The overriding principle is that when making a PR, you shouldn't touch anything not directly related to the PR. Unfortunately I have limited time to review bigrquery PRs, so anything you can do to reduce the amount of code I need to review helps.

File: tests/testthat/test-bq-perform.R [hunk 10]
```diff
@@ -85,4 +86,84 @@ test_that("can supply parameters", {
   df <- bq_table_download(job_tb)
   expect_equal(nrow(df), 1)
   expect_equal(df$phase, "New Moon")
+
+})
+
+test_that("can supply parameters as array for IN statement", {
+
+  query_template <-
```
Comment (hadley) [inline tests/testthat/test-bq-perform.R:94]: I think you could simplify this by using SQL like `SELECT ARRAY_LENGTH(@x) AS length`

**Comment by hadley** [comment]
I've simplified the tests (while maintaining coverage) and tweaked the docs. @byapparov Please take a look and let me know if you see any problems.

File: R/bq-param.R [hunk 11]
```diff
@@ -1,25 +1,31 @@
-
-
+#' Explicitly define queriy parameter
```
Comment (jennybc) [inline R/bq-param.R]:
```suggestion
#' Explicitly define query parameter
```

File: R/bq-param.R [hunk 12]
```diff
@@ -1,25 +1,31 @@
...
+#'
+#' By default, bigrquery will assume vectors of length 1 are scalars,
+#' and longer vectors are arrays. If you need to pass a length-1 array,
+#' you'll need to explicitly use `bq_param_arary()`.
```
Comment (jennybc) [inline R/bq-param.R:5]:
```suggestion
#' you'll need to explicitly use `bq_param_array()`.
```

**Comment by byapparov** [comment]
@hadley In general the test change you made for bq_array_param is fine. the only thing, there were cases to test that it works correctly for: 
1. scalar variable, 
2. if array is passed as vector (without wrapping into bq_array_param)

I don't think these tests are completely redundant. 

However, the case you left is a preferred way of passing parameters, so I think it is good.

**Comment by hadley** [comment]
@byapparov those are handled in the `bq_params` tests, so I don't think it's important to re-test on the `bq_perform_query()` side.
